opal/datatype/opal_datatype_pack.c

```diff
@@ -111,155 +108,123 @@ opal_pack_homogeneous_contig_with_gaps_function( opal_convertor_t* pConv,
                                                  uint32_t* out_size,
                                                  size_t* max_data )
 {
+    size_t remaining, length, initial_bytes_converted = pConv->bConverted;
     const opal_datatype_t* pData = pConv->pDesc;
     dt_stack_t* stack = pConv->pStack;
+    ptrdiff_t extent = pData->ub - pData->lb;
     unsigned char *user_memory, *packed_buffer;
-    uint32_t iov_count, index;
+    uint32_t idx = 0;
     size_t i;
-    size_t bConverted, remaining, length, initial_bytes_converted = pConv->bConverted;
-    ptrdiff_t extent= pData->ub - pData->lb;
-    ptrdiff_t initial_displ = pConv->use_desc->desc[pConv->use_desc->used].end_loop.first_elem_disp;
 
+    /* The memory layout is contiguous with gaps in the begining and at the end. The datatype true_lb
+     * is the initial displacement, the size the length of the contiguous area and the extent represent
+     * how much we should jump between elements.
+     */
     assert( (pData->flags & OPAL_DATATYPE_FLAG_CONTIGUOUS) && ((ptrdiff_t)pData->size != extent) );
     DO_DEBUG( opal_output( 0, "pack_homogeneous_contig( pBaseBuf %p, iov_count %d )\n",
                            (void*)pConv->pBaseBuf, *out_size ); );
     if( stack[1].type != opal_datatype_uint1.id ) {
         stack[1].count *= opal_datatype_basicDatatypes[stack[1].type]->size;
         stack[1].type = opal_datatype_uint1.id;
     }
+    /* We can provide directly the pointers in the user buffers (like the convertor_raw) */
+    if( NULL == iov[0].iov_base ) {
+        user_memory = pConv->pBaseBuf + pData->true_lb + stack[0].disp + stack[1].disp;
```

Comment:
why not 
- moving this line between the if {} and the for {} (after line # 144)
- removing line 142

Reply:
It seems possible to rewrite this code without the ```if``` but I am not sure if it is really improves the readability or performance. Let me give it a try.